pkg/sbom/collectors/containerd/containerd.go

```diff
@@ -20,104 +20,147 @@ import (
+	containerdScanRequest, ok := request.(ScanRequest)
 	if !ok {
-		return sbom.ScanResult{Error: fmt.Errorf("invalid request type '%s' for collector '%s'", reflect.TypeOf(request), collectorName)}
+		return sbom.ScanResult{Error: fmt.Errorf("invalid request type '%s' for collector '%s'", reflect.TypeOf(request), collectors.ContainerdCollector)}
+	}
+
+	if c.containerdClient == nil {
+		cl, err := cutil.NewContainerdUtil()
+		if err != nil {
+			return sbom.ScanResult{Error: fmt.Errorf("error creating containerd client: %s", err)}
+		}
+		c.containerdClient = cl
+	}
+
+	wlm := workloadmeta.GetGlobalStore()
+	if wlm == nil {
+		return sbom.ScanResult{Error: fmt.Errorf("workloadmeta store is not initialized")}
 	}
+	imageMeta, err := wlm.GetImage(containerdScanRequest.ID())
+	if err != nil {
+		return sbom.ScanResult{Error: fmt.Errorf("image metadata not found for image id %s: %s", containerdScanRequest.ID(), err)}
+	}
+	log.Infof("containerd scan request [%v]: scanning image %v", containerdScanRequest.ID(), imageMeta.Name)
 
-	if containerdScanRequest.ImageMeta != nil {
-		log.Infof("containerd scan request [%v]: scanning image %v", containerdScanRequest.ID(), containerdScanRequest.ImageMeta.Name)
+	image, err := c.containerdClient.Image(imageMeta.Namespace, imageMeta.Name)
+	if err != nil {
+		return sbom.ScanResult{Error: fmt.Errorf("error getting image %s/%s: %s", imageMeta.Namespace, imageMeta.Name, err)}
 	}
 
 	var report sbom.Report
-	var err error
-	if containerdScanRequest.FromFilesystem {
+	if c.fromFileSystem {
 		report, err = c.trivyCollector.ScanContainerdImageFromFilesystem(
 			ctx,
-			containerdScanRequest.ImageMeta,
-			containerdScanRequest.Image,
-			containerdScanRequest.ContainerdClient,
-			opts,
+			imageMeta,
+			image,
+			c.containerdClient,
+			c.opts,
 		)
 	} else {
 		report, err = c.trivyCollector.ScanContainerdImage(
 			ctx,
-			containerdScanRequest.ImageMeta,
-			containerdScanRequest.Image,
-			containerdScanRequest.ContainerdClient,
-			opts,
+			imageMeta,
+			image,
+			c.containerdClient,
+			c.opts,
 		)
 	}
 	scanResult := sbom.ScanResult{
 		Error:   err,
 		Report:  report,
-		ImgMeta: containerdScanRequest.ImageMeta,
+		ImgMeta: imageMeta,
 	}
 
 	return scanResult
 }
 
+// Type returns the container image scan type
+func (c *Collector) Type() collectors.ScanType {
+	return collectors.ContainerImageScanType
+}
+
+// Channel returns the channel to send scan results
+func (c *Collector) Channel() chan sbom.ScanResult {
+	return c.resChan
+}
+
+// Options returns the collector options
+func (c *Collector) Options() sbom.ScanOptions {
+	return c.opts
+}
+
+// Shutdown shuts down the collector
+func (c *Collector) Shutdown() {
+	if c.resChan != nil && !c.closed {
+		close(c.resChan)
+	}
+	c.closed = true
+}
+
 func init() {
-	collectors.RegisterCollector(collectorName, &Collector{})
+	collectors.RegisterCollector(collectors.ContainerdCollector, &Collector{
+		resChan: make(chan sbom.ScanResult, 1000),
```

Comment:
It shouldn't matter that much given that 2000 is already a lot. When an agent starts, it scans all the images, sometimes multiple times if the image has multiple RepoDigests/RepoTags. The workqueue should allow us to avoid scanning an image more than once.